_SPIKE_.md

```diff
@@ -0,0 +1,496 @@
+  public inclusionResolvers: {[key: string]: InclusionResolver};
+
+  // ...
+}
+```
+
+**IMPORTANT:** To support use cases where no custom repository class is created
+and applications/extensions are instantiating `DefaultCrudRepository` class
+directly, it's important to expose `inclusionResolvers` as a public property.
+
+When a repository method like `find`, `findOne` and `findById` is called to
+query the database, it should use use `inclusionResolvers` map to fetch any
+related models requested by `filter.include`.
+
+Conceptually:
+
+```ts
+export class DefaultCrudRepository<T, Relations> {
+  // ...
+
+  async find(
+    filter?: Filter<T>,
+    options?: Options,
+  ): Promise<(T & Relations)[]> {
+    const models = await ensurePromise(
+      this.modelClass.find(this.normalizeFilter(filter), options),
+    );
+    const entities = this.toEntities(models);
+    // NEW LINE ADDED TO RESOLVE INCLUDED MODELS:
+    return this.includeRelatedModels(entities, filter, options);
+  }
+
+  protected async includeRelatedModels(
+    entities: T[],
+    filter?: Filter<T>,
+    options?: Options,
+  ): Promise<(T & Relations)[]> {
+    const result = entities as (T & Relations)[];
+
+    // process relations in parallel
+    const resolveTasks = filter.include.map(i => {
+      const relationName = i.relation;
+      const handler = this.inclusionResolvers[relationName];
+      return handler.fetchIncludedModels(entities, i, options);
+    });
+    await Promise.all(resolveTasks);
+
+    return result;
+  }
+}
+```
+
+### 3. Model-specific repositories register inclusion resolvers
+
+Model-specific repository classes (e.g. `CategoryRepository`) should register
+inclusion resolvers for model relations, similarly to how we are creating
+relation-repository factories now.
+
+Conceptually:
+
+```ts
+export class CategoryRepository extends DefaultCrudRepository {
+  products: HasManyRepositoryFactory<Product, typeof Category.prototype.id>;
+
+  constructor(
+    dataSource: juggler.DataSource,
+    productRepositoryGetter: Getter<ProductRepository>,
+  ) {
+    super(Category, dataSource);
+
+    // we already have this line to create HasManyRepository factory
+    this.products = this.createHasManyRepositoryFactoryFor(
+      'todos',
+      productRepositoryGetter,
+    );
+
+    // add this line to register inclusion resolver
+    this.registerHasManyInclusion('products', this.productRepositoryGetter);
+  }
+}
+```
+
+As we realized in LB3, not all relations are allowed to be traversed. For
+example, when fetching users, we don't want the callers to include related
+`AccessToken` models in the response!
+
+In the proposed solution, this is solved by NOT REGISTERING any inclusion
+resolver for such relation.
+
+In the future, we can implement a "dummy" resolver that will report a helpful
+error for such relations (rather than a generic "unknown inclusion" error).
+
+```ts
+// usage
+this.registerForbiddenInclusion('accessTokens');
```

Comment:
Maybe it's better to name it as something like `this.prohibitInclusion('accessTokens')`?

Reply:
Done in 507820f